out_writeapi_test.go

```diff
@@ -38,3 +44,143 @@ func TestFLBPluginRegister(t *testing.T) {
+	return m.managedstream.Close()
+}
+
+func TestFLBPluginFlushCtx(t *testing.T) {
+	var checks StreamChecks
+	var setID interface{}
+	var id int
+
+	mockMS := &MockManagedStream{
+		AppendRowsFunc: func(ctx context.Context, data [][]byte, opts ...managedwriter.AppendOption) (*managedwriter.AppendResult, error) {
+			checks.appendRows = true
+			return nil, nil
+		},
+	}
+
+	originalFunc := getWriter
+	getWriter = func(client *managedwriter.Client, ctx context.Context, projectID string, opts ...managedwriter.WriterOption) (MWManagedStream, error) {
+		return mockMS, nil
+	}
+	defer func() { getWriter = originalFunc }()
+
+	patch1 := monkey.Patch(output.FLBPluginConfigKey, func(plugin unsafe.Pointer, key string) string {
+		return ""
+	})
+	defer patch1.Unpatch()
+
+	patchDescriptor := monkey.Patch(getDescriptors, func(curr_ctx context.Context, managed_writer_client *managedwriter.Client, project string, dataset string, table string) (protoreflect.MessageDescriptor, *descriptorpb.DescriptorProto) {
+		return nil, nil
+	})
+	defer patchDescriptor.Unpatch()
+
+	patchSetContext := monkey.Patch(output.FLBPluginSetContext, func(plugin unsafe.Pointer, ctx interface{}) {
+		log.Println("Ran set context")
+		setID = ctx
+	})
+	defer patchSetContext.Unpatch()
+
+	plugin := unsafe.Pointer(nil)
+	// ctx := unsafe.Pointer(nil)
+	initRes := FLBPluginInit(plugin)
+
+	orgFunc := getContext
+	getContext = func(ctx unsafe.Pointer) int {
+		checks.calledGetContext = true
+		uintptrValueBack := uintptr(ctx)
+		id = int(uintptrValueBack)
+		return id
+	}
+	defer func() { getContext = orgFunc }()
+
+	patchCheckResponses := monkey.Patch(checkResponses, func(curr_ctx context.Context, currQueuePointer *[]*managedwriter.AppendResult, waitForResponse bool) int {
+		checks.calledcheckResponses = true
+		return 0
+	})
+	defer patchCheckResponses.Unpatch()
+
+	patchDecoder := monkey.Patch(output.NewDecoder, func(data unsafe.Pointer, length int) *output.FLBDecoder {
+		checks.createDecoder = true
+		return nil
+	})
+	defer patchDecoder.Unpatch()
+
+	var loopCount int = 0
+	patchRecord := monkey.Patch(output.GetRecord, func(dec *output.FLBDecoder) (ret int, ts interface{}, rec map[interface{}]interface{}) {
+		checks.gotRecord = true
+		if loopCount%2 == 0 {
+			loopCount++
+			return 0, nil, nil
+		} else {
+			loopCount++
+			return 1, nil, nil
+		}
+	})
+	defer patchRecord.Unpatch()
+
+	patchParse := monkey.Patch(parseMap, func(mapInterface map[interface{}]interface{}) map[string]interface{} {
+		checks.calledparseMap = true
+		return nil
+	})
+	defer patchParse.Unpatch()
+
+	patchJTB := monkey.Patch(json_to_binary, func(message_descriptor protoreflect.MessageDescriptor, jsonRow map[string]interface{}) ([]byte, error) {
+		checks.calledJTB = true
+		tempBD := make([]byte, 1)
+		tempBD[0] = 1
+		return tempBD, nil
+	})
+	defer patchJTB.Unpatch()
+
+	// Gets config so we can check length of initial results queue
+	config := configMap[id]
```

Comment:
configMap is an internal data structure used by the plugin, so it should not be tested in this way. Perhaps you can add a test-only method that retrieves the queue length given an id.

Reply:
Would this be an internal function on the plugin that I can just call in the test file too? Because I remember Seth and I were looking into it and Godoesn't have any specific infrastructure to only activate a method during testing.

Reply:
Yes it would be an function on the plugin. That way it is not dependent on the specific implementation details, but allows you to inspect the queue length. No need to worry about mutex locking in its implementation as this is for testing purposes only.

Reply:
Done.